### PR TITLE
Lecture 2 plp demodata pdp

### PR DESCRIPTION
**Pull request by thatglance**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    public void init() throws ServletException {
+        super.init();
+        dao = ArrayListProductDao.getInstance();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String uri = request.getRequestURI();
+        int lastSlashIndex = uri.lastIndexOf("/");
+        String id = uri.substring(lastSlashIndex + 1);
```
Comment (nikifarau) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: param validation

File: src/main/java/com/es/phoneshop/web/SampleDataServletContextListener.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import javax.servlet.ServletContextListener;
+import java.math.BigDecimal;
+import java.util.Currency;
+
+public class SampleDataServletContextListener implements ServletContextListener {
+
+    @Override
+    public void contextInitialized(ServletContextEvent servletContextEvent) {
+        String insertSampleData = servletContextEvent.getServletContext().getInitParameter("insertSampleData");
+        if (insertSampleData != null && Boolean.valueOf(insertSampleData)) {
```
Comment (nikifarau) [inline src/main/java/com/es/phoneshop/web/SampleDataServletContextListener.java]: Boolean.valueOf is enough

File: src/main/java/com/es/phoneshop/web/SampleDataServletContextListener.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            dao.save(new Product(9L, "nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
+            dao.save(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
+            dao.save(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
+            dao.save(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
+            dao.save(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
+        }
+    }
+
+    @Override
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
```
Comment (AntonNikifarau) [inline src/main/java/com/es/phoneshop/web/SampleDataServletContextListener.java]: useless overriding method

File: src/main/webapp/WEB-INF/pages/error.jsp [hunk 4]
```diff
@@ -0,0 +1,24 @@
+<%@ page contentType="text/html;charset=UTF-8" %>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+
+<html>
+<head>
+    <title>Product List</title>
+    <link href='http://fonts.googleapis.com/css?family=Lobster+Two' rel='stylesheet' type='text/css'>
+    <link rel="stylesheet" href="${pageContext.servletContext.contextPath}/styles/main.css">
```
Comment (nikifarau) [inline src/main/webapp/WEB-INF/pages/error.jsp]: jsp:include

File: src/main/webapp/WEB-INF/pages/productList.jsp [hunk 5]
```diff
@@ -20,20 +20,34 @@
...
+      </form>
       <table>
         <thead>
           <tr>
             <td>Image</td>
-            <td>Description</td>
-            <td class="price">Price</td>
+            <td>
+              Description
+              <a href="${pageContext.servletContext.contextPath}/products?sortField=description&sortMode=asc&query=${param.query}">asc</a>
```
Comment (nikifarau) [inline src/main/webapp/WEB-INF/pages/productList.jsp]: c:url

File: src/main/webapp/WEB-INF/web.xml [hunk 6]
```diff
@@ -15,4 +24,18 @@
...
+    <servlet-class>com.es.phoneshop.web.ProductDetailsPageServlet</servlet-class>
+  </servlet>
+
+  <servlet-mapping>
+    <servlet-name>productDetails</servlet-name>
+    <url-pattern>/products/*</url-pattern>
+  </servlet-mapping>
+
+  <error-page>
+    <error-code>500</error-code>
```
Comment (nikifarau) [inline src/main/webapp/WEB-INF/web.xml]: product not found case is not handled

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 7]
```diff
@@ -1,22 +1,80 @@
...
 import java.util.Optional;
 import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
-    private List<Product> productsList = new ArrayList<>();
+
+    private static volatile ArrayListProductDao instance;
+    private List<Product> productsList;
+
+    private ArrayListProductDao() {
```
Comment (nikifarau) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:14]: unit tests

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 8]
```diff
@@ -1,22 +1,80 @@
...
         return productsList.stream().filter(p -> p.getId().equals(id)).findFirst().orElse(null);
     }
 
     @Override
-    public synchronized List<Product> findProducts() {
-        return productsList.stream().filter(p -> (p.getPrice() != null) && (p.getStock() > 0)).
-                collect(Collectors.toList());
+    public synchronized List<Product> findProducts(final String query, final String sortField, final String sortMode) {
+        return productsList.stream().filter(p -> (p.getPrice() != null) && (p.getStock() > 0))
+                .filter(p -> {
```
Comment (nikifarau) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: make it more readable :)
hints:
comparator = comparator.thenComparing(comparator2).reversed();
filter = predicat.and(predicat2);
productsList.stream().filter(filter).sorted(comparator).collect(toList());

**Pull request merged**